pkg/cmd/builder.go

```diff
@@ -0,0 +1,198 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/spf13/pflag"
+	apimeta "k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"github.com/kubernetes-incubator/custom-metrics-apiserver/pkg/cmd/server"
+	"github.com/kubernetes-incubator/custom-metrics-apiserver/pkg/dynamicmapper"
+	"github.com/kubernetes-incubator/custom-metrics-apiserver/pkg/provider"
+)
+
+type AdapterBase struct {
+	*server.CustomMetricsAdapterServerOptions
+
+	// Name is the name of the API server.  It defaults to custom-metrics-adapter
+	Name string
+
+	// RemoteKubeConfigFile specifies the kubeconfig to use to construct
+	// the dynamic client and RESTMapper.  It's set from a flag.
+	RemoteKubeConfigFile string
+	// DiscoveryInterval specifies the interval at which to recheck discovery
+	// information for the discovery RESTMapper.  It's set from a flag.
+	DiscoveryInterval time.Duration
```

Comment:
If you're interested, it's the thing in kubernetes that maps kinds to resources.  The distinction is a bit tricky, since it deals with how kubernetes APIs work under the hood a bit, but essentially, Kinds represent types in the language and API, and resources represent locations/URL structures.  They're often similar-looking, but not always.

For instance,  the `apps/v1.Deployment` Kind directly maps to the `apps/v1.deployments` Resource (which corresponds to the `/apis/apps/v1/namespaces/<ns>/deployments` URL structure), but also is used by the `apps/v1.deployments/status` subresource (subresources are just a special type of resoure) (which corresponds to the `/apis/apps/v1/namespaces/<ns>/deployments/<dep>/status` URL structure).

Alternatively, the `autoscaling/v1.Scale` kind doesn't directly correspond to any resources, and instead is used by most (not all, but that's another story) up-to-date `/scale` subresources.

We use the RESTMapper to figure out what the Kind is from a given URL fragment, and vice versa.  So when you say `jobs.batch`, we say `Kind: Jobs, Group: Batch` in the `describedObject` field, and know how to go a list all the jobs on the cluster based on RESTMapper information.

Reply:
thanks for the explanation